tools/f18/f18-parse-demo.cc

```diff
@@ -0,0 +1,476 @@
+// Copyright (c) 2018-2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// F18 parsing demonstration.
+//   f18-parse-demo [ -E | -fdump-parse-tree | -funparse-only ]
+//     foo.{f,F,f77,F77,f90,F90,&c.}
+//
+// By default, runs the supplied source files through the F18 preprocessing and
+// parsing phases, reconstitutes a Fortran program from the parse tree, and
+// passes that Fortran program to a Fortran compiler identified by the $F18_FC
+// environment variable (defaulting to gfortran).  The Fortran preprocessor is
+// always run, whatever the case of the source file extension.  Unrecognized
+// options are passed through to the underlying Fortran compiler.
+//
+// This program is actually a stripped-down variant of f18.cc, a temporary
+// scaffolding compiler driver that can test some semantic passes of the
+// F18 compiler under development.
+
+#include "../../lib/common/default-kinds.h"
+#include "../../lib/parser/characters.h"
+#include "../../lib/parser/features.h"
+#include "../../lib/parser/message.h"
+#include "../../lib/parser/parse-tree-visitor.h"
+#include "../../lib/parser/parse-tree.h"
+#include "../../lib/parser/parsing.h"
+#include "../../lib/parser/provenance.h"
+#include "../../lib/parser/unparse.h"
+#include "../../lib/semantics/dump-parse-tree.h"
```

Comment:
It would make sense for this tool to only depend on parser and not semantics.

I guess ParseTreeDumper is in semantics because if a Name has a Symbol, that gets dumped. We could decouple ParseTreeDumper from semantics by passing in a Name dumping function. The version of that function for this tool would just print the name.

Reply:
Right.  Does anybody use the parse tree dumper?  Maybe it just doesn't need the symbol dump.

Reply:
I use the parse tree dumper and I think it's useful for understanding the parse tree.
I would be okay with removing the symbol dump part. Mostly I use `-fdebug-dump-symbols` anyway.

Reply:
I'll try moving the parse tree dumper into the parser proper.

Reply:
Done.